modules/zip/src/filesystems/zip-filesystem.ts

```diff
@@ -0,0 +1,132 @@
+import {FileSystem, isBrowser} from '@loaders.gl/core';
+import {FileProvider, isFileProvider} from '../file-provider/file-provider';
+import {FileHandleFile} from '../file-provider/file-handle-file';
+import {ZipCDFileHeader, zipCDFileHeaderGenerator} from '../parse-zip/cd-file-header';
+import {parseZipLocalFileHeader} from '../parse-zip/local-file-header';
+
+type CompressionHandler = (compressedFile: ArrayBuffer) => Promise<ArrayBuffer>;
+/** Handling different compression types in zip */
+const COMPRESSION_METHODS: {[key: number]: CompressionHandler} = {
+  /** No compression */
+  0: async (compressedFile) => compressedFile
+};
+
+/**
+ * FileSystem adapter for a ZIP file
+ * Holds FileProvider object that provides random access to archived files
+ */
+export class ZipFileSystem implements FileSystem {
+  /** FileProvider instance promise */
+  private fileProvider: Promise<FileProvider | null> = Promise.resolve(null);
+
+  /**
+   * Constructor
+   * @param file - instance of FileProvider or file path string
+   */
+  constructor(file: FileProvider | string) {
+    // Try to open file in NodeJS
+    if (typeof file === 'string') {
+      if (!isBrowser) {
+        this.fileProvider = FileHandleFile.from(file);
+      } else {
+        throw new Error('Cannot open file for random access in a WEB browser');
```

Comment:
Nit: You can if you use the `File` reader trick I showed you. That would presumably go into a FIleProvider and not change this class.